tsconfig.json

```diff
@@ -1,16 +1,12 @@
 {
   "compilerOptions": {
     "esModuleInterop": true,
-    "exactOptionalPropertyTypes": true,
-    "forceConsistentCasingInFileNames": true,
     "lib": ["ES2020"],
     "module": "CommonJS",
     "moduleResolution": "node",
     "noEmit": true,
-    "noErrorTruncation": true,
-    "noUncheckedIndexedAccess": true,
```

Comment:
This seems like a downgrade to type safety.

Reply:
Hi, thanks for checking in on this! Since these flags are not enabled in the core monorepo's tsconfig, it was necessary to verify ahead of migration that disabling them won't break anything in this package. 

My understanding is that `exactOptionalPropertyTypes` is in the process of being enabled in the core monorepo. Any further improvements in type safety will also propagate to this package once it's been migrated.

Reply:
Why can't this be done in the core repository itself? I don't think we should be merging this.

Reply:
This repo will be archived as soon as the `eth-json-rpc-provider` package is migrated, and all future development will take place in the core monorepo. #27 and #28 are preparatory steps for that process. These links might provide more context:
https://github.com/MetaMask/core/issues/1551
https://github.com/MetaMask/core/pull/1707#issuecomment-1733311813

Reply:
1. @legobeat None of the non-root packages in the core monorepo seem to have project-specific compiler options. Is the monorepo set up to accommodate them without manual adjustments to compiler/linter/test settings? Also, is adding project-specific config in alignment with the intended direction of the monorepo? This is my first ticket, and I'm not entirely familiar with the build system or coding guidelines yet.

2. There are a number of deps where this package is ahead of the core repo. I don't know if it makes sense to wait for them to be updated upstream, when any version bumps in the core repo will be propagated to the `eth-json-rpc-provider` subrepo once it's migrated. It doesn't seem to me like there are any long-term concerns to consider here, since this repo will be archived as soon as the migrated subrepo is merged, and that should happen asap.

Reply:
We can consider upgrading TypeScript in the monorepo rather than downgrading it here. Not sure why I hadn't thought of that, good suggestion.

Adding additional compiler options might be more difficult though. We can explore that too, but I recall trying to enable `noUncheckedIndexedAccess` once before and getting stuck on a type error caused by a library we use. That was a while ago though, we should try again.

But enabling these flags on a per-package basis might be a lot easier to get working, and provide a path to incrementally adding flags in the future. I haven't tried doing that before.